### PR TITLE
Custom entries are superglobal

### LINKED ISSUE #13654: Error "Undeclared custom entry: dfrac" when Exporting module
#### Description of the problem

```coq
Module dfrac.
  Declare Custom Entry dfrac.
  Notation "{ dq }" := (dq) (in custom dfrac at level 1, dq constr).
End dfrac.

Module M1.
  Export dfrac.
  Notation "↦∗ dq" := (id dq) (at level 20, dq custom dfrac at level 1).
End M1.

Module M2.
  Export dfrac.
  Notation "$∗ dq" := (id dq) (at level 20, dq custom dfrac at level 1).
End M2.

Module M.
  Export M1.
  Export M2.
  (* Warning: Custom entry dfrac has been overriden. [custom-entry-overriden,parsing] *)
End M.

Import M.
(* Error: Undeclared custom entry: dfrac. *)
```

The module `dfrac` defines a custom entry, and both the modules `M1` and `M2` define a notion that involves the custom entry. Now, exporting both `M1` and `M2` is fine. However, when importing the module `M` that exports said modules, I get the error "Undeclared custom entry: dfrac".

Also, there is a warning when exporting `M2` in `M`. I don't understand that warning.

This code is obtained by reducing an actual bug we faced when trying to use custom entries in Iris, see https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/554#note_61015.

```coq
From iris.heap_lang Require proofmode.

Module foo.
  Export iris.heap_lang.derived_laws.
  Export iris.heap_lang.primitive_laws.
End foo.

(* Undeclared custom entry: dfrac. *)
Export foo.
```

This requires the version of Iris in https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/554

#### Coq Version

The Coq Proof Assistant, version 8.12.0 (September 2020)
compiled on Sep 15 2020 7:57:30 with OCaml 4.08.1

Probably not relevant, but that's on Debian Linux.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #13654

Overlay: https://github.com/DeepSpec/sfdev/pull/327 (backwards compatible, on secret repo)

### CODE REVIEW

**Comment by ppedrot** [comment]
I agree that in the current setting, this is the proper fix. But shouldn't we also ponder the mid-term possibility to give well-scoped names to custom entries?

File: vernac/metasyntax.ml [hunk 1]
```diff
@@ -1862,15 +1862,11 @@ let warn_custom_entry =
          (fun s ->
           strbrk "Custom entry " ++ str s ++ strbrk " has been overridden.")
 
-let load_custom_entry _ _ = ()
-
-let open_custom_entry _ (_,(local,s)) =
+let load_custom_entry _ (_,(local,s)) =
   if Egramcoq.exists_custom_entry s then warn_custom_entry s
```
Comment (ppedrot) [inline vernac/metasyntax.ml:1866]: Isn't this going to trigger a lot of warning in case of diamonds?

Comment (SkySkimmer) [reply vernac/metasyntax.ml:1866]: Not sure what you mean, example?

Comment (ppedrot) [reply vernac/metasyntax.ml:1866]: If you require a library defining a custom entry twice you'll get a warning. Before, you had to import it which is harder to do involuntarily, except maybe for `Export` statements. There is the same kind of problem with all superglobal objects btw.

Comment (SkySkimmer) [reply vernac/metasyntax.ml:1866]: I believe redundant requires are skipped, so no warning.

Comment (SkySkimmer) [reply vernac/metasyntax.ml:1866]: Isn't it correct to warn if the entry is defined twice?
If you meant requiring twice AFAIK the second require is skipped.

**Comment by herbelin** [comment]
Thanks a lot @SkySkimmer.

**Comment by charguer** [comment]
I have concrete examples involving `Require` in diamonds. Please let me know when there is a branch ready for me to test. Thanks

**Comment by SkySkimmer** [comment]
This should be ready to test AFAICT

**Comment by ppedrot** [comment]
@SkySkimmer what about the test-suite failures?

**Comment by SkySkimmer** [comment]
The test suite tests that custom entries follow import and so needs changing. I'll do it later today.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 2]
```diff
@@ -997,6 +997,9 @@ Custom entries
    This command supports the :attr:`local` attribute, which limits the entry to the
    current module.
 
+   When not local, custom entries survive module closing and are
+   declared when a file is Required.
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1001]:
```suggestion
   Non-local custom entries survive module closing and are
   declared when a file is Required.
```

File: doc/changelog/03-notations/14183-custom-load.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+- **Fixed:**
+  Custom entries follow Require, not Import
```
Comment (jfehrle) [inline doc/changelog/03-notations/14183-custom-load.rst:2]:
```suggestion
  Non-local custom entries survive module closing and are
  declared when a file is Required.
```

It was a bit too terse.

**Review by jfehrle — COMMENTED** [review]
Wording suggestions

**Comment by ppedrot** [comment]
@SkySkimmer what about the doc tweaks?

**Comment by herbelin** [comment]
>  But shouldn't we also ponder the mid-term possibility to give well-scoped names to custom entries?

Meaning here qualified names? If yes, this may raise a few incompatibilities when the module where a `Declare Scope scope` occurs is not opened, right?

Anyway, to get qualified names, here is a rough analysis of what would need to be done at first view:
- in `Notation`, use a `KNmap` instead of `String.Map` for `scope_map`; use `None` rather than `""` for the `default_scope`
- in `Metasyntax`, add a `module ScopeTab = Nametab.Make(FullPath)(KerName)` and (at first view) a similar business as in e.g. `tacenv.ml` (from `push_tactic`, renamed as `push_scope`, to `shortest_qualid_of_tactic` renamed as `shortest_qualid_of_scope`, again at first view)
- in `Metasyntax.declare_scope` and following, use `add_leaf` instead with the scope name turned into an identifier
-  in `load_scope_command_common` uses the `kn` instead of `scope` for `declare_scope` and add `ScopeTab.push_scope (Nametab.Until i)` in `load` and `ScopeTab.push_scope (Nametab.Exactly i)` in `open`, `ScopeTab.push_scope (Nametab.Until 1)` in `cache_md`

For compatibility, accept temporarily to get a scope by basename even if visible only by qualified name.

**Comment by SkySkimmer** [comment]
@herbelin wouldn't that be for scopes, not custom entries?

**Comment by herbelin** [comment]
Oups, yes. This collided with a previous request I had in mind about scopes. For custom entries, this would be similar with `load_custom_entry` and cie, if wanted. Anyway, my comment was mostly to suggest that this kind of qualification is a priori possible to implement if there some will for it. No more ambition.

**Comment by ppedrot** [comment]
@herbelin like all libobjects that can depend on other ones, this is problematic for functors and interfaces though. You would be able to replay the issue of #14002 by replacing databases with entries and hints with notations.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Merging soon.

**Pull request merged**